### DISCUSSION TITLE
ReDesign the life cycle of Molecule instantiation

### DISCUSSION THREAD

**Discussion by wewoor** [DTStack/molecule#675] [Ideas]
1. 控制 Extension 的执行时机

**Comment by wewoor** [comment]
我目前设想 API 类似如下的调用：

```ts
import { create, Workbench } from '@molecule/ide';

import '@dtinsight/molecule/esm/style/mo.css';

const moInstance = create({
    defaultLocale: '',
    extensions: [],
}, );

const App = () => moInstance.render(<Workbench />);

ReactDOM.render(<App />, document.getElementById('root'));

```

其他 API 调用：

```ts
// Inactive specify extension
moInstance.extension.inactive(() => ());
// activityBar add
moInstance.activityBar.add();
```

**Reply by wewoor** [reply]
@mortalYoung @kiwiwong

**Comment by mortalYoung** [comment]
<img width="1197" alt="image" src="https://user-images.githubusercontent.com/18719701/156751708-0ef8c273-4299-4971-864b-f9e4803142c7.png">

**Comment by mortalYoung** [comment]
# RFC: Re-design the life cycle of molecule instantiation

## Goals
- Ensure the orders among locale, colorTheme, common extensions are correct
- To make Services registed more clearly

## Background
We met the life cycle problems several times. And the `provider/molecule` file became more and more fragile. This file was filled with comments so we could notice the one who edit this file next time why we did this trick magic in here. And it's time to refactor it.

## Proposal
In order to let molecule to be a standalone application, we can have a **standaloneService** so that we can have a service to control the count of instantiation. And we also need a **instantiationService** to control the whole life cycle of current instantiation. And we can define the orders of Services. We still export a function called **create**,  so user could call **create** function to get the instantiation of molecule, and everything belongs to this instantiation.
```js
const moInstance = create({
    defaultLocale: '',
    extensions: [],
}, );

const App = () => moInstance.render(<Workbench />);

ReactDOM.render(<App />, document.getElementById('root'));
```
Besides, we still export the **Provider** to compatible with the older usage of molecule.

**⚠️ Notice: Although we will try to make the breaking changes more less, there are still some breaking changes unavoidable.**
